File: game/xp.py

def get_xp_required(level):
    """
    Возвращает количество требуемых очков для поднятия указанного уровня
    """
    return 30 * (level - 1) ** 1.3


def get_xp_info(points):
    """
    Возвращает информацию о рейтинге игрока:
    - текущий уровень
    - текущее количество очков внутри уровня
    - количество очков, необходимое для повышеия уровня
    """
    level = 0
    points_needed = get_xp_required(level + 1)
    
    while points >= points_needed:
        points = points - points_needed
        level = level + 1
        points_needed = get_xp_required(level + 1)

    return {
        'level': level,
        'points_current': int(points),
        'points_needed': int(points_needed),
    }

File: frontend/src/js/shared/xp.js
import { byId } from './utils';

/**
 * Загружает и выводит мнформацию о рейтинге и уровне пользователя на верхнюю панель
 */
export function displayXpInfo() {
    fetch('/get_xp_info/').then(response => {
        return new Promise((resolve) => {
            if (response.status !== 200) throw response.status;
            resolve(response.json());
        });
    }).then(data => {
        byId('level').innerText = data.level;
        byId('xp-bar').setAttribute('title', `Очков: ${data['points_current']} / ${data['points_needed']}`);

        const perc = data['points_current'] / data['points_needed'] * 100;
        byId('xp-bar__filled').style.width = `${perc}%`;
        byId('xp-bar__empty').style.width = `${100 - perc}%`;
    }).catch(() => {});
}
